File: shared-plugins/adjust/android/AdjustPlugin.java
// TODO: Change this package to match your project
package com.yourpackage;

import android.util.Log;

import com.adjust.sdk.Adjust;
import com.adjust.sdk.AdjustAdRevenue;
import com.adjust.sdk.AdjustConfig;
import com.adjust.sdk.AdjustEvent;
import com.adjust.sdk.LogLevel;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.util.Iterator;

/**
 * Custom Capacitor plugin for Adjust attribution SDK. Initializes with
 * environment-aware config, tracks typed events with optional callback
 * parameters and revenue, and emits attribution changes to JS.
 */
@CapacitorPlugin(name = "AdjustPlugin")
public class AdjustPlugin extends Plugin {
    private static final String TAG = "AdjustPlugin";
    private boolean isInitialized = false;
    private boolean isDebugMode = false;

    @PluginMethod
    public void initialize(PluginCall call) {
        String appToken = call.getString("appToken");
        String environment = call.getString("environment", "sandbox");
        String logLevel = call.getString("logLevel", "verbose");

        if (appToken == null || appToken.isEmpty()) {
            call.reject("App token is required");
            return;
        }

        try {
            String adjustEnvironment = environment.equals("production") 
                ? AdjustConfig.ENVIRONMENT_PRODUCTION 
                : AdjustConfig.ENVIRONMENT_SANDBOX;

            AdjustConfig config = new AdjustConfig(getContext(), appToken, adjustEnvironment);
            
            isDebugMode = environment.equals("sandbox");
            switch (logLevel) {
                case "verbose":
                    config.setLogLevel(LogLevel.VERBOSE);
                    break;
                case "debug":
                    config.setLogLevel(LogLevel.DEBUG);
                    break;
                case "info":
                    config.setLogLevel(LogLevel.INFO);
                    break;
                case "warn":
                    config.setLogLevel(LogLevel.WARN);
                    break;
                case "error":
                    config.setLogLevel(LogLevel.ERROR);
                    break;
                case "suppress":
                    config.setLogLevel(LogLevel.SUPRESS);
                    isDebugMode = false;
                    break;
                default:
                    config.setLogLevel(LogLevel.VERBOSE);
            }

            config.setOnAttributionChangedListener(attribution -> {
                if (isDebugMode) {
                    Log.d(TAG, "Attribution changed");
                }
                JSObject data = new JSObject();
                data.put("trackerToken", attribution.trackerToken);
                data.put("trackerName", attribution.trackerName);
                data.put("network", attribution.network);
                data.put("campaign", attribution.campaign);
                data.put("adgroup", attribution.adgroup);
                data.put("creative", attribution.creative);
                data.put("clickLabel", attribution.clickLabel);
                notifyListeners("attributionChanged", data);
            });

            Adjust.initSdk(config);
            isInitialized = true;

            if (isDebugMode) {
                Log.d(TAG, "Adjust SDK initialized successfully");
            }
            
            JSObject result = new JSObject();
            result.put("initialized", true);
            call.resolve(result);
        } catch (Exception e) {
            if (isDebugMode) {
                Log.e(TAG, "Failed to initialize Adjust SDK", e);
            }
            call.reject("Failed to initialize Adjust SDK");
        }
    }

    @PluginMethod
    public void trackEvent(PluginCall call) {
        if (!isInitialized) {
            call.reject("Adjust SDK not initialized");
            return;
        }

        String eventToken = call.getString("eventToken");
        if (eventToken == null || eventToken.isEmpty()) {
            call.reject("Event token is required");
            return;
        }

        try {
            AdjustEvent event = new AdjustEvent(eventToken);

            JSObject callbackParams = call.getObject("callbackParameters");
            if (callbackParams != null) {
                Iterator<String> callbackKeys = callbackParams.keys();
                while (callbackKeys.hasNext()) {
                    String key = callbackKeys.next();
                    event.addCallbackParameter(key, callbackParams.getString(key));
                }
            }

            JSObject partnerParams = call.getObject("partnerParameters");
            if (partnerParams != null) {
                Iterator<String> partnerKeys = partnerParams.keys();
                while (partnerKeys.hasNext()) {
                    String key = partnerKeys.next();
                    event.addPartnerParameter(key, partnerParams.getString(key));
                }
            }

            Double revenue = call.getDouble("revenue");
            String currency = call.getString("currency");
            if (revenue != null && currency != null) {
                event.setRevenue(revenue, currency);
            }

            Adjust.trackEvent(event);
            
            if (isDebugMode) {
                Log.d(TAG, "Event tracked");
            }
            call.resolve();
        } catch (Exception e) {
            if (isDebugMode) {
                Log.e(TAG, "Failed to track event", e);
            }
            call.reject("Failed to track event");
        }
    }

    @PluginMethod
    public void trackAdRevenue(PluginCall call) {
        if (!isInitialized) {
            call.reject("Adjust SDK not initialized");
            return;
        }

        String source = call.getString("source", "admob_sdk");
        Double revenue = call.getDouble("revenue");
        String currency = call.getString("currency", "USD");

        if (revenue == null) {
            call.reject("Revenue is required");
            return;
        }

        try {
            AdjustAdRevenue adRevenue = new AdjustAdRevenue(source);
            adRevenue.setRevenue(revenue, currency);

            Integer adImpressionsCount = call.getInt("adImpressionsCount");
            if (adImpressionsCount != null) {
                adRevenue.setAdImpressionsCount(adImpressionsCount);
            }

            String adRevenueNetwork = call.getString("adRevenueNetwork");
            if (adRevenueNetwork != null) {
                adRevenue.setAdRevenueNetwork(adRevenueNetwork);
            }

            String adRevenueUnit = call.getString("adRevenueUnit");
            if (adRevenueUnit != null) {
                adRevenue.setAdRevenueUnit(adRevenueUnit);
            }

            String adRevenuePlacement = call.getString("adRevenuePlacement");
            if (adRevenuePlacement != null) {
                adRevenue.setAdRevenuePlacement(adRevenuePlacement);
            }

            Adjust.trackAdRevenue(adRevenue);
            
            if (isDebugMode) {
                Log.d(TAG, "Ad revenue tracked");
            }
            call.resolve();
        } catch (Exception e) {
            if (isDebugMode) {
                Log.e(TAG, "Failed to track ad revenue", e);
            }
            call.reject("Failed to track ad revenue");
        }
    }

    @PluginMethod
    public void isEnabled(PluginCall call) {
        JSObject result = new JSObject();
        result.put("enabled", isInitialized);
        call.resolve(result);
    }

    @PluginMethod
    public void setEnabled(PluginCall call) {
        Boolean enabled = call.getBoolean("enabled", true);
        
        if (enabled) {
            Adjust.enable();
        } else {
            Adjust.disable();
        }
        
        call.resolve();
    }

    @PluginMethod
    public void gdprForgetMe(PluginCall call) {
        Adjust.gdprForgetMe(getContext());
        call.resolve();
    }
}

File: shared-plugins/adjust/ios/AdjustPlugin.swift
import Foundation
import Capacitor
import AdjustSdk

/// Custom Capacitor plugin for Adjust attribution SDK. Initializes with
/// environment-aware config, tracks typed events, and provides an attribution
/// change delegate.
@objc(AdjustPlugin)
public class AdjustPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "AdjustPlugin"
    public let jsName = "AdjustPlugin"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "trackEvent", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "trackAdRevenue", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isEnabled", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setEnabled", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "gdprForgetMe", returnType: CAPPluginReturnPromise),
    ]
    private var isInitialized = false

    @objc func initialize(_ call: CAPPluginCall) {
        guard let appToken = call.getString("appToken"), !appToken.isEmpty else {
            call.reject("App token is required")
            return
        }

        let environment = call.getString("environment") ?? "sandbox"
        let logLevel = call.getString("logLevel") ?? "verbose"

        let adjustEnvironment = environment == "production"
            ? ADJEnvironmentProduction
            : ADJEnvironmentSandbox

        guard let config = ADJConfig(appToken: appToken, environment: adjustEnvironment) else {
            call.reject("Failed to create Adjust config")
            return
        }

        switch logLevel {
        case "verbose": config.logLevel = .verbose
        case "debug":   config.logLevel = .debug
        case "info":    config.logLevel = .info
        case "warn":    config.logLevel = .warn
        case "error":   config.logLevel = .error
        case "suppress": config.logLevel = .suppress
        default: config.logLevel = .verbose
        }

        if let waitInterval = call.getInt("attConsentWaitingInterval"), waitInterval > 0 {
            config.attConsentWaitingInterval = UInt(waitInterval)
        }

        config.delegate = self

        Adjust.initSdk(config)
        isInitialized = true

        call.resolve(["initialized": true])
    }

    @objc func trackEvent(_ call: CAPPluginCall) {
        guard isInitialized else {
            call.reject("Adjust SDK not initialized")
            return
        }

        guard let eventToken = call.getString("eventToken"), !eventToken.isEmpty else {
            call.reject("Event token is required")
            return
        }

        guard let event = ADJEvent(eventToken: eventToken) else {
            call.reject("Failed to create event")
            return
        }

        if let callbackParams = call.getObject("callbackParameters") {
            for (key, value) in callbackParams {
                if let stringValue = value as? String {
                    event.addCallbackParameter(key, value: stringValue)
                }
            }
        }

        if let partnerParams = call.getObject("partnerParameters") {
            for (key, value) in partnerParams {
                if let stringValue = value as? String {
                    event.addPartnerParameter(key, value: stringValue)
                }
            }
        }

        if let revenue = call.getDouble("revenue"), let currency = call.getString("currency") {
            event.setRevenue(revenue, currency: currency)
        }

        Adjust.trackEvent(event)
        call.resolve()
    }

    @objc func trackAdRevenue(_ call: CAPPluginCall) {
        guard isInitialized else {
            call.reject("Adjust SDK not initialized")
            return
        }

        guard let revenue = call.getDouble("revenue") else {
            call.reject("Revenue is required")
            return
        }

        let source = call.getString("source") ?? "admob_sdk"
        let currency = call.getString("currency") ?? "USD"

        guard let adRevenue = ADJAdRevenue(source: source) else {
            call.reject("Failed to create ad revenue object")
            return
        }
        adRevenue.setRevenue(revenue, currency: currency)

        if let count = call.getInt("adImpressionsCount") {
            adRevenue.setAdImpressionsCount(Int32(count))
        }
        if let network = call.getString("adRevenueNetwork") {
            adRevenue.setAdRevenueNetwork(network)
        }
        if let unit = call.getString("adRevenueUnit") {
            adRevenue.setAdRevenueUnit(unit)
        }
        if let placement = call.getString("adRevenuePlacement") {
            adRevenue.setAdRevenuePlacement(placement)
        }

        Adjust.trackAdRevenue(adRevenue)
        call.resolve()
    }

    @objc func isEnabled(_ call: CAPPluginCall) {
        call.resolve(["enabled": isInitialized])
    }

    @objc func setEnabled(_ call: CAPPluginCall) {
        let enabled = call.getBool("enabled") ?? true
        if enabled {
            Adjust.enable()
        } else {
            Adjust.disable()
        }
        call.resolve()
    }

    @objc func gdprForgetMe(_ call: CAPPluginCall) {
        Adjust.gdprForgetMe()
        call.resolve()
    }
}

extension AdjustPlugin: AdjustDelegate {
    public func adjustAttributionChanged(_ attribution: ADJAttribution?) {
        guard let attribution = attribution else { return }
        notifyListeners("attributionChanged", data: [
            "trackerToken": attribution.trackerToken ?? "",
            "trackerName": attribution.trackerName ?? "",
            "network": attribution.network ?? "",
            "campaign": attribution.campaign ?? "",
            "adgroup": attribution.adgroup ?? "",
            "creative": attribution.creative ?? "",
            "clickLabel": attribution.clickLabel ?? ""
        ])
    }
}
